Update the condition to write timestamp attributes

robertasg opened the pull request:
### Summary

Saving a record in a transaction that fails resets the changed attributes.

If the given record has timestamp recording enabled, saving the same instance of the record again results in a connection adapter's `ConstraintException`.

That happens because the partial write doesn't include the timestamp fields, as their state was reset after the failed transaction.

In other words, this fixes https://github.com/rails/rails/issues/33443

twitnithegirl commented on activerecord/test/cases/timestamp_test.rb:
```diff
@@ -456,6 +456,17 @@ def test_all_timestamp_attributes_in_model
+    end
+
+    assert_nothing_raised { toy.save! }
+  end
```
This test is failing for mysql2 as follows:
```
TimestampTest#test_creating_a_record_after_a_failed_transaction:
ActiveRecord::NotNullViolation: Mysql2::Error: Field 'created_at' doesn't have a default value: INSERT INTO `toys` VALUES ()
    /home/travis/.rvm/gems/ruby-2.4.4/gems/mysql2-0.5.2/lib/mysql2/client.rb:131:in `_query'
    /home/travis/.rvm/gems/ruby-2.4.4/gems/mysql2-0.5.2/lib/mysql2/client.rb:131:in `block in query'
    /home/travis/.rvm/gems/ruby-2.4.4/gems/mysql2-0.5.2/lib/mysql2/client.rb:130:in `handle_interrupt'
    /home/travis/.rvm/gems/ruby-2.4.4/gems/mysql2-0.5.2/lib/mysql2/client.rb:130:in `query'
    /home/travis/build/rails/rails/activerecord/lib/active_record/connection_adapters/abstract_mysql_adapter.rb:185:in `block (2 levels) in execute'
    /home/travis/build/rails/rails/activesupport/lib/active_support/dependencies/interlock.rb:48:in `block in permit_concurrent_loads'
    /home/travis/build/rails/rails/activesupport/lib/active_support/concurrency/share_lock.rb:187:in `yield_shares'
    /home/travis/build/rails/rails/activesupport/lib/active_support/dependencies/interlock.rb:47:in `permit_concurrent_loads'
    /home/travis/build/rails/rails/activerecord/lib/active_record/connection_adapters/abstract_mysql_adapter.rb:184:in `block in execute'
    /home/travis/build/rails/rails/activerecord/lib/active_record/connection_adapters/abstract_adapter.rb:584:in `block (2 levels) in log'
    /home/travis/.rvm/rubies/ruby-2.4.4/lib/ruby/2.4.0/monitor.rb:214:in `mon_synchronize'
    /home/travis/build/rails/rails/activerecord/lib/active_record/connection_adapters/abstract_adapter.rb:583:in `block in log'
    /home/travis/build/rails/rails/activesupport/lib/active_support/notifications/instrumenter.rb:23:in `instrument'
    /home/travis/build/rails/rails/activerecord/lib/active_record/connection_adapters/abstract_adapter.rb:574:in `log'
    /home/travis/build/rails/rails/activerecord/lib/active_record/connection_adapters/abstract_mysql_adapter.rb:183:in `execute'
    /home/travis/build/rails/rails/activerecord/lib/active_record/connection_adapters/mysql/database_statements.rb:28:in `execute'
    /home/travis/build/rails/rails/activerecord/lib/active_record/connection_adapters/abstract_mysql_adapter.rb:194:in `execute_and_free'
    /home/travis/build/rails/rails/activerecord/lib/active_record/connection_adapters/mysql/database_statements.rb:33:in `exec_query'
    /home/travis/build/rails/rails/activerecord/lib/active_record/connection_adapters/abstract/database_statements.rb:117:in `exec_insert'
    /home/travis/build/rails/rails/activerecord/lib/active_record/connection_adapters/abstract/database_statements.rb:149:in `insert'
    /home/travis/build/rails/rails/activerecord/lib/active_record/connection_adapters/abstract/query_cache.rb:21:in `insert'
    /home/travis/build/rails/rails/activerecord/lib/active_record/persistence.rb:197:in `_insert_record'
    /home/travis/build/rails/rails/activerecord/lib/active_record/persistence.rb:745:in `_create_record'
    /home/travis/build/rails/rails/activerecord/lib/active_record/counter_cache.rb:184:in `_create_record'
    /home/travis/build/rails/rails/activerecord/lib/active_record/locking/optimistic.rb:70:in `_create_record'
    /home/travis/build/rails/rails/activerecord/lib/active_record/attribute_methods/dirty.rb:175:in `_create_record'
    /home/travis/build/rails/rails/activerecord/lib/active_record/callbacks.rb:332:in `block in _create_record'
    /home/travis/build/rails/rails/activesupport/lib/active_support/callbacks.rb:132:in `run_callbacks'
    /home/travis/build/rails/rails/activesupport/lib/active_support/callbacks.rb:816:in `_run_create_callbacks'
    /home/travis/build/rails/rails/activerecord/lib/active_record/callbacks.rb:332:in `_create_record'
    /home/travis/build/rails/rails/activerecord/lib/active_record/timestamp.rb:100:in `_create_record'
    /home/travis/build/rails/rails/activerecord/lib/active_record/persistence.rb:716:in `create_or_update'
    /home/travis/build/rails/rails/activerecord/lib/active_record/callbacks.rb:328:in `block in create_or_update'
    /home/travis/build/rails/rails/activesupport/lib/active_support/callbacks.rb:132:in `run_callbacks'
    /home/travis/build/rails/rails/activesupport/lib/active_support/callbacks.rb:816:in `_run_save_callbacks'
    /home/travis/build/rails/rails/activerecord/lib/active_record/callbacks.rb:328:in `create_or_update'
    /home/travis/build/rails/rails/activerecord/lib/active_record/persistence.rb:318:in `save!'
    /home/travis/build/rails/rails/activerecord/lib/active_record/validations.rb:52:in `save!'
    /home/travis/build/rails/rails/activerecord/lib/active_record/transactions.rb:313:in `block in save!'
    /home/travis/build/rails/rails/activerecord/lib/active_record/transactions.rb:374:in `block in with_transaction_returning_status'
    /home/travis/build/rails/rails/activerecord/lib/active_record/connection_adapters/abstract/database_statements.rb:254:in `block in transaction'
    /home/travis/build/rails/rails/activerecord/lib/active_record/connection_adapters/abstract/transaction.rb:238:in `block in within_new_transaction'
    /home/travis/.rvm/rubies/ruby-2.4.4/lib/ruby/2.4.0/monitor.rb:214:in `mon_synchronize'
    /home/travis/build/rails/rails/activerecord/lib/active_record/connection_adapters/abstract/transaction.rb:235:in `within_new_transaction'
    /home/travis/build/rails/rails/activerecord/lib/active_record/connection_adapters/abstract/database_statements.rb:254:in `transaction'
    /home/travis/build/rails/rails/activerecord/lib/active_record/transactions.rb:212:in `transaction'
    /home/travis/build/rails/rails/activerecord/lib/active_record/transactions.rb:372:in `with_transaction_returning_status'
    /home/travis/build/rails/rails/activerecord/lib/active_record/transactions.rb:313:in `save!'
    /home/travis/build/rails/rails/activerecord/lib/active_record/suppressor.rb:48:in `save!'
    /home/travis/build/rails/rails/activerecord/test/cases/timestamp_test.rb:468:in `block in test_creating_a_record_after_a_failed_transaction'
    /home/travis/build/rails/rails/activesupport/lib/active_support/testing/assertions.rb:32:in `assert_nothing_raised'
    /home/travis/build/rails/rails/activerecord/test/cases/timestamp_test.rb:468:in `test_creating_a_record_after_a_failed_transaction'
rails test home/travis/build/rails/rails/activerecord/test/cases/timestamp_test.rb:460
```

twitnithegirl replied:
These should help:
- https://optimize-this.blogspot.com/2012/04/datetime-default-now-finally-available.html
- https://solidfoundationwebdev.com/blog/posts/default-model-attributes-in-rails

robertasg replied:
Seems that this uncovers an issues that's related to `ActiveRecord::Dirty` and the `mysql2` adapter. I've added a workaround to fix the reported issue.

Commit:
Update the condition to write timestamp attributes

Write timestamp attributes when creating a record unless the attributes were already changed.
